sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -229,6 +229,24 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper {
     !plan.isStreaming && hasSelectivePredicate(plan)
   }
 
+  /**
+   * Check the partition has static filter. So we can skip add dynamic partition pruning
+   * if there already exists.
+   */
+  private def hasStaticPartitionEqualityCondition(e: Expression, p: LogicalPlan): Boolean = {
+    p.find {
```

Comment:
Shall we simply look at `LogicalPlan.constraints`? It's a bit risky to find a `Filter` in an arbitrary position.

Reply:
but the one issue is that if people disable `constraintPropagation`, the `LogicalPlan.constraints` will not work.

Reply:
OK. How about a different approach: we add DPP filters anyway, and then remove unnecessary ones in `CleanupDynamicPruningFilters`.

In that rule, the predicates are already pushed dow to the source, so we can simply look at the predicates above scan, and remove DPP filters that has a corresponding static partition filters.

Reply:
looks good, and add rule `BooleanSimplification` behind `CleanupDynamicPruningFilters` to cleanup the true predicates.